src/fault_log.rs

```diff
@@ -0,0 +1,87 @@
+//! Logging functionality for reporting faulty node behavior.
+//!
+//! The subalgorithms (CommonCoin, Agreement, Broadcast, CommonSubset,
+//! SyncKeyGen) propogate their faulty node logs upwards to the top-level
+//! algorithms (HoneyBadger, DynamicHoneyBadger) via DistAlgorihm`'s
+//! `.input()` and `.handle_message()` trait methods. All faulty node logs
+//! are then aggregated together by the top-level algorithms and stored
+//! within their structures as a `fault_log` field.
+
+/// Represents each reason why a Node could be considered faulty.
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum FaultKind {
+    /// CommonCoin received a signature share from an unverified sender.
+    UnverifiedSignatureShareSender,
+    /// HoneyBadger received a decryption share from an unverified sender.
+    UnverifiedDecryptionShareSender,
+    /// HoneyBadger was unable to deserialize a proposer's ciphertext.
+    InvalidCiphertext,
+    /// HoneyBadger was unable to decrypt a share received from a proposer.
+    ShareDecryptionFailed,
+    /// Broadcast recevied an Echo message containing an invalid proof.
+    InvalidProof,
+    /// HoneyBadger could not deserialze bytes (ie. a serialized Batch)
+    /// from a given proposer into a vector of transactions.
+    BatchDeserializationFailed,
+    /// DynamicHoneyBadger received a message (Accept, Propose, or Change)
+    /// with an invalid signature.
+    IncorrectPayloadSignature,
+    /// DynamicHoneyBadger/SyncKeyGen received an invalid Accept message.
+    InvalidAcceptMessage,
+    /// DynamicHoneyBadger/SyncKeyGen received an invalid Propose message.
+    InvalidProposeMessage,
+}
+
+/// A structure representing the context of a faulty Node. This structure
+/// describes which Node is faulty (`node_id`) and which faulty behavior
+/// that the node exhibited ('kind').
+#[derive(Clone)]
+pub struct Fault<NodeUid: Clone> {
+    pub node_id: NodeUid,
+    pub kind: FaultKind,
+}
+
+impl<NodeUid: Clone> Fault<NodeUid> {
+    pub fn new(node_id: &NodeUid, kind: FaultKind) -> Self {
```

Comment:
I passed by reference because most (sub)algorithm methods that call `FaultLog::new()` already had a `NodeUid` reference, so I guess I did it to cut down on clones. If you think it's  unidiomatic (I don't necessarily think that it is), it's no problem to change it.

Reply:
(It _is_ unidiomatic: https://rust-lang-nursery.github.io/api-guidelines/flexibility.html#caller-decides-where-to-copy-and-place-data-c-caller-control — but I'm still fine with the exception, too. Your call.)